README.md
# skflex

![GitHub](https://img.shields.io/github/license/ltw85/skflex) ![PyPI](https://img.shields.io/pypi/v/skflex) [![Build Status](https://scrutinizer-ci.com/g/LTW85/skflex/badges/build.png?b=main)](https://scrutinizer-ci.com/g/LTW85/skflex/build-status/main) [![CodeFactor](https://www.codefactor.io/repository/github/ltw85/skflex/badge)](https://www.codefactor.io/repository/github/ltw85/skflex) [![Tweet](https://img.shields.io/twitter/url?style=social&url=https%3A%2F%2Fgithub.com%2FLTW85%2Fskflex)](https://twitter.com/intent/tweet?text=https%3A%2F%2Fgithub.com%2FLTW85%2Fskflex)

# *FLEXIBLE FUNCTIONS* ----- *FAST PROCESSING AND EVALUATION*

skflex provides a suite of utility functions for use with the sklearn library. The module primarily focuses on producing typical plots and metrics for evaluating machine learning models. It has been designed with flexability and customisation in mind to speed up workflows, and enhance comparative evaluation. 

# Installation and Import
```
pip install skflex

import skflex.skflex as skf
```
_____________________________________________________________________________________________________________________________________________________________________

# ROC_AUC Curve

### roc_auc_plot(*models, X_test = None, y_test = None, width = 14, height = 12, legend_size = 14, title='ROC Curve'*)

Accepts fitted model(s) and test data. It will then:
- Calculate ROC
- Calculate AUC
- Plot ROC curve with AUC provided in the legend

#### Parameters:
- models - fitted model objects. NOTE: Only models with a 'predict_proba' or 'decision_function' method are supported.
- X_test - test feature set
- y_test - test labels set
- title - title for ROC curve plot
- width - plot width
- height - plot height
- legend_size - size of plot legend

#### Example:

```
from sklearn.naive_bayes import GaussianNB
from sklearn.linear_model import LogisticRegression

model_1 = GaussianNB()
model_2 = LogisticRegression()

model_1.fit(X_train, y_train)
model_2.fit(X_train, y_train)

skf.roc_auc_plot(model_1, model_2, X_test = X_test, y_test = y_test, 
                title = 'Example ROC plot')
```

![roc plot](example_roc_plot.PNG)

_________________________________________________________________________________________________________________________________________________________________________________________________________________________

# Classification Reports

### classifier_train_report(*models, X_train = None, y_train = None, X_test = None, y_test = None, scoring = 'accuracy', title = 'Reports'*)

Accepts classifier models, training data, and test data. It will then:
- Fit the model(s) to training data
- Make predictions using test data
- Produce classification report for comparison
- Produce confusion matrix for comparison
- Provide an ordered summary (ranked best to worst score) using given evaluation metric

#### Parameters:
- models - model objects to be trained and evaluated
- X_train - training feature set
- y_train - training label set
- X_test - test feature set
- y_test - test label set
- scoring - summary evaluation metric. Common classifier evaluation metrics including accuracy, f1, precision, and recall are supported. Refer to [sklean scoring documentation](https://scikit-learn.org/stable/modules/model_evaluation.html#accuracy-score) for more information. Scoring methodologies should be passed as strings, for example, precision would be passed as `scoring = 'precision'` 
- title - title for output report 

#### Example:

```
from sklearn.naive_bayes import GaussianNB
from sklearn.linear_model import LogisticRegression

model_1 = GaussianNB()
model_2 = LogisticRegression()

skf.classifier_train_report(model_1, model_2, X_train = X_train, y_train = y_train, 
                            X_test = X_test, 
                            y_test = y_test, 
                            scoring = 'accuracy', 
                            title = 'Example Report')
```

![classification report](example_classification_report1.PNG)

_________________________________________________________________________________________________________________________________________________________________________________________________________________________

# Validation Curve

### validation_plot(*model = None, param = None, param_grid = None, X_train = None, y_train = None, cv = 5, scoring = 'accuracy', width = 9, height = 9, title = 'Validation Curve'*)

Accepts a model, a related hyper-parameter, a list of hyper-parameter values, training and test data, number of cross-validation folds, scoring methodology, as well as a plot title.
It will produce a plot of the validation curve for the training and test data using the mean scores and standard deviations obtained through the cross-validation process. 

#### Parameters:
- model - model object 
- param - hyperparameter to be used to produce the validation curve 
- param_grid - hyperparameter values to be tested
- X_train - training feature set
- y_train - training label set
- cv - number of cross-validation folds
- scoring - scoring methodology used during cross-validation process
- title - title for validation plot
- width - plot width
- height - plot height

#### Example:

```
from sklearn.tree import DecisionTreeClassifier

model_1 = DecisionTreeClassifier()
params = [5, 10, 15, 20, 30, 40, 50]

skf.validation_plot(model = model_1, param = 'max_depth', param_grid = params, 
                    X_train = X_train, 
                    y_train = y_train, 
                    title = 'Example Validation Curve')
```

![validation plot](example_validation_curve.PNG)

____________________________________________________________________________________________________________________________________________________________________________________________________________________________

# Train, Validation and Test Split

### train_val_test(*data = None, class_labels = None, train = 0.6, val = 0.2, shuffle = True, random_state = None*)

Accepts a Pandas dataframe and will return a training, validation, and test set. Operates in a similar fashion to the sklearn train_test_split function by defining a percentage split for the training and validation sets (example 0.6 = 60%). The remainder is allocated to the test set.

#### Parameters:
- data - dataframe to be split into a training, validation, and test set
- class_labels - column in the dataframe containing class labels
- train - percentage of data to be allocated to the training set
- val - percentage of data to be allocated to the validation set
- shuffle - if true, will shuffle the rows in the dataframe before splitting
- random_state - if shuffle is ture, will set a random seed so that ordering produced by shuffle can be reproduced

Returns: X_train, y_train, X_val, y_val, X_test, y_test

#### Example:

```
X_train, y_train, X_val, y_val, X_test, y_test = skf.train_val_test(data = my_data, 
                                                                    class_labels = 'labels', 
                                                                    train = 0.6, 
                                                                    val = 0.2)
```

______________________________________________________________________________________________________________________________________________________________________________________________________________________________

# PCA Scree Plot

### pca_scree_plot(*data = None, n_components = None, width = 16, height = 10, legend_size = 12, scale_data = False, title = 'PCA Scree Plot'*)

Accepts data (array/dataframe), and number of principal components to be analysed. It will produce a scree plot of the cumulative variance explained.

#### Parameters:
- data - dataset to be analysed
- n_components - number of principal components to be analysed 
- width - width of plot
- height - height of plot
- legend_size - size of plot legend
- scale_data - normalises data before analysis and plotting. If the data being passed has not yet been normalised, this parameter should be set as `scale_data = True`
- title - plot title

### Example:

```
from sklearn.preprocessing import scale

n_data = scale(my_data)

skf.pca_scree_plot(data = n_data, n_components = 70, title = 'Example PCA Scree Plot')
```

![scree plot](example_scree_plot.PNG)

_______________________________________________________________________________________________________________________________________________________________________________________________________________________________________

## Contributions

Contributions of any kind and pull requests welcome!

_______________________________________________________________________________________________________________________________________________________________________________________________________________________________________

## Requirements
- Sklearn >= 0.24.1
- Matplotlib >= 3.3.4
- Pandas >= 1.2.4
- Numpy >= 1.20.1


CHANGELOG.txt
Change Log
============

0.0.1 (16/10/2021)

- First release.

1.0.0 (18/10/2021)

- Fixed bug with scoring parameter for classifier_train_report. The function now accepts the strings 'accuracy', 'f1', 'recall' and 'precision' as arguments for scoring.

- added a 'scale_data' parameter for the pca_scree_plot function that will normalise data passed in before performing analysis and plotting.

- Update docs and provide examples

1.0.1 (19/10/2021)

- Fixed bug with y axis label for validation plot

- added further explanation for classifier_train_report summary

1.0.2 (19/10/2021)

- Update docs

sklearn_utils.py
from sklearn.metrics import roc_curve
from sklearn.metrics import roc_auc_score
from sklearn.model_selection import validation_curve
from sklearn.decomposition import PCA
import matplotlib
import matplotlib.pyplot as plt
from sklearn.metrics import classification_report
from sklearn.metrics import confusion_matrix
import sklearn.metrics
import numpy as np
import pandas as pd

def roc_auc_plot(*models, X_test = None, y_test = None, width = 14, height = 12, legend_size = 14, title='ROC Curve'):
   
    """
    Function that accepts fitted model(s) and test data. It will then:
    - Calculate ROC
    - Calculate AUC
    - Plot ROC curve with AUC provided in the legend

    """

    rndm_probs = [0 for _ in range(len(y_test))]
    rndm_auc = roc_auc_score(y_test, rndm_probs)
    rndm_fpr, rndm_tpr, _ = roc_curve(y_test, rndm_probs)

    plt.subplots(1, figsize=(width, height))
    plt.plot(rndm_fpr, rndm_tpr, linestyle='--', label='Random Chance - AUC = %.1f' % (rndm_auc))
    plt.xlabel('False Positive Rate')
    plt.ylabel('True Positive Rate')
    plt.title(title, fontsize=16, fontweight='bold')
   
    
    for i in models:
        model_name = type(i).__name__
        
        if hasattr( i, 'predict_proba' ) and callable( i.predict_proba ):
            probs = i.predict_proba(X_test)
            probs = probs[: , 1]
        else:
            probs = i.decision_function(X_test)
            
        auc = roc_auc_score(y_test, probs)
        fpr, tpr, _ = roc_curve(y_test, probs)
        plt.plot(fpr, tpr, marker='.', label= model_name + ' - AUC = %.4f' % (auc))
        
    plt.legend(loc = 'lower right', prop={'size': legend_size})  
    plt.show()

def classifier_train_report(*models, training_data_X = None, training_data_y = None, test_data_X = None, test_data_y = None, scoring = accuracy_score, title = 'Reports'):
    
    """
    function that accepts classifier models, training data, and test data. It will then:
    - Fit the model(s) to training data
    - Make predictions using test data
    - Produce classification report for comparison
    - Produce confusion matrix for comparison
    
    """
    
    print('~'*50 + title + '~'*50)
    
    m_scores = []
    for i in models:
        model_name = type(i).__name__
        i.fit(training_data_X, training_data_y)
        y_pred = i.predict(test_data_X)
        print()
        print('-'*20 + model_name + ' ' + 'Classification Report' + '-'*20)
        print(classification_report(y_pred, test_data_y)) 
        print()
        print('-'*20 + model_name + ' ' + 'Confusion Matrix' + '-'*20)
        print(confusion_matrix(y_pred, test_data_y))
        print()
        print('*'*100) 
        print()
        score = scoring(test_data_y, y_pred)
        m_scores.append({'model': model_name, 'score': score})
        
    print('Summary:') 
    for s in m_scores:
        print('model:' + ' ' + str(s['model']) + ' ' + '--- ' + 'score:' + ' ' + str(s['score'].round(3)))   

def validation_plot(model = None, param = None, param_grid = None, X_train = None, y_train = None, cv = 5, scoring = 'accuracy', width = 9, height = 9, title = 'Validation Curve'):

    """ 
    Function that accepts a model, a related hyper-parameter, a list of hyper-parameter values, training and test data, number of cross-validation folds, scoreing methodology, as well as a plot title.
    It will produce a plot of the validation curve for the training and test data using the mean scores and standard deviations obtained through the cross-validation process.  
    
    """
    train_scores, test_scores = validation_curve(
    model, X_train, y_train, param_name=param, param_range=param_grid,
    scoring=scoring, cv=cv)
    
    train_mean = np.mean(train_scores, axis=1)
    train_std = np.std(train_scores, axis=1)

    test_mean = np.mean(test_scores, axis=1)
    test_std = np.std(test_scores, axis=1)
    
    plt.subplots(1, figsize = (width, height))
    plt.plot(param_grid, train_mean, label = 'Training score', color = 'black')
    plt.plot(param_grid, test_mean, label = 'Validation score', color = 'brown')

    plt.fill_between(param_grid, train_mean - train_std, train_mean + train_std, color = 'blue', alpha = 0.2)
    plt.fill_between(param_grid, test_mean - test_std, test_mean + test_std, color = 'darkblue', alpha = 0.2)
 
    plt.title(title, fontsize=14, fontweight='bold')
    plt.xlabel('Param Range')
    plt.ylabel('Accuracy Score')
    plt.tight_layout()
    plt.legend(loc = 'best')
    plt.show()

def train_val_test(data = None, class_labels = None, train = 0.6, val = 0.2, shuffle = True, random_state = None):
    
    """
    Function that accepts a Pandas dataframe and will return a training, validation, and test set. 
    """
    
    if shuffle:
        data = data.sample(frac = 1, random_state = random_state)
    else:
        pass
    
    X = data.drop(class_labels, axis = 1)
    y = pd.DataFrame(data[class_labels])
    
    split = train + val
    X_train, X_val, X_test = np.split(X, [int(train*len(X)), int(split*len(X))])
    y_train, y_val, y_test = np.split(y, [int(train*len(y)), int(split*len(y))])
    
    return X_train, y_train, X_val, y_val, X_test, y_test

def pca_scree_plot(data = None, n_components = None, width = 16, height = 10, legend_size = 12, title = 'PCA Scree Plot'):

    """
    Function that accepts data, and number of principal components to be analysed. It will produce a scree plot of the cumulative variance explained.  
    """
    
    pca = PCA(n_components = n_components)
    pca_model = pca.fit(data)
    var_exp = pca_model.explained_variance_ratio_.cumsum().round(4)*100
    
    fig, ax = plt.subplots(figsize=(width, height))
    ax.set_title(title, fontsize=14, fontweight='bold')
    ax.set_xlabel('Principal Components', fontsize=12)
    ax.set_ylabel('Variance Explained (%)', fontsize=12)
    ax.axhspan(90, 100, alpha = 0.3, color = '#FF8C78', label = '90% - 95%')
    ax.axhspan(95, 100, alpha = 0.5, color = '#FF8C78', label = '95% - 99%')
    ax.axhspan(99, 100, alpha = 0.7, color = '#FF8C78', label = '99% - 100%')
    ax.legend(loc = 4, prop={'size': legend_size})
    ax.plot(var_exp, marker = '.', markersize = 10);
    plt.show()    

skflex.py
from sklearn.metrics import roc_curve
from sklearn.metrics import roc_auc_score
from sklearn.model_selection import validation_curve
from sklearn.decomposition import PCA
from sklearn.preprocessing import scale
import matplotlib
import matplotlib.pyplot as plt
from sklearn.metrics import classification_report
from sklearn.metrics import confusion_matrix
from sklearn.metrics import accuracy_score
from sklearn.metrics import f1_score
from sklearn.metrics import precision_score
from sklearn.metrics import recall_score
import numpy as np
import pandas as pd

def roc_auc_plot(*models, X_test = None, y_test = None, width = 14, height = 12, legend_size = 14, title = 'ROC Curve'):
   
    """
    Function that accepts fitted model(s) and test data. It will then:
    - Calculate ROC
    - Calculate AUC
    - Plot ROC curve with AUC provided in the legend

    """

    rndm_probs = [0 for _ in range(len(y_test))]
    rndm_auc = roc_auc_score(y_test, rndm_probs)
    rndm_fpr, rndm_tpr, _ = roc_curve(y_test, rndm_probs)

    plt.subplots(1, figsize=(width, height))
    plt.plot(rndm_fpr, rndm_tpr, linestyle='--', label='Random Chance - AUC = %.1f' % (rndm_auc))
    plt.xlabel('False Positive Rate')
    plt.ylabel('True Positive Rate')
    plt.title(title, fontsize=16, fontweight='bold')
   
    
    for i in models:
        model_name = type(i).__name__
        
        if hasattr( i, 'predict_proba' ) and callable( i.predict_proba ):
            probs = i.predict_proba(X_test)
            probs = probs[: , 1]
        else:
            probs = i.decision_function(X_test)
            
        auc = roc_auc_score(y_test, probs)
        fpr, tpr, _ = roc_curve(y_test, probs)
        plt.plot(fpr, tpr, marker = '.', label= model_name + ' - AUC = %.4f' % (auc))
        
    plt.legend(loc = 'lower right', prop={'size': legend_size}) 
    plt.tight_layout() 
    plt.show()

def classifier_train_report(*models, X_train = None, y_train = None, X_test = None, y_test = None, scoring = 'accuracy', title = 'Reports'):
    
    """
    function that accepts classifier models, training data, and test data. It will then:
    - Fit the model(s) to training data
    - Make predictions using test data
    - Produce classification report for comparison
    - Produce confusion matrix for comparison
    - Provide ordered summary using given evaluation metric
    
    """
    
    print('~'*50 + title + '~'*50)
    
    m_scores = []

    for i in models:
        model_name = type(i).__name__
        i.fit(X_train, y_train)
        y_pred = i.predict(X_test)
        print()
        print('-'*20 + model_name + ' ' + 'Classification Report' + '-'*20)
        print(classification_report(y_pred, y_test)) 
        print()
        print('-'*20 + model_name + ' ' + 'Confusion Matrix' + '-'*20)
        print(confusion_matrix(y_pred, y_test))
        print()
        print('*'*100) 
        print()
        
        if scoring == 'accuracy':
            sum_score = accuracy_score(y_test, y_pred)
        elif scoring == 'f1':
            sum_score = f1_score(y_test, y_pred)
        elif scoring == 'precision':
            sum_score = precision_score(y_test, y_pred)
        elif scoring == 'recall':
            sum_score = recall_score(y_test, y_pred)

        m_scores.append({'model': model_name, 'sum_score': sum_score})
        
    sorted_scores = sorted(m_scores, key=lambda s: s['sum_score'], reverse = True)
    
    print('SUMMARY - models sorted by' + ' ' + str(scoring) + ' ' + 'score:') 
    for s in sorted_scores:
        print('model:' + ' ' + str(s['model']) + ' ' + '--- ' + 'score:' + ' ' + str(s['sum_score'].round(3)))   

def validation_plot(model = None, param = None, param_grid = None, X_train = None, y_train = None, cv = 5, scoring = 'accuracy', width = 9, height = 9, 
                    title = 'Validation Curve'):

    """ 
    Function that accepts a model, a related hyper-parameter, a list of hyper-parameter values, training and test data, number of cross-validation folds, scoreing methodology, as well as a plot title.
    It will produce a plot of the validation curve for the training and test data using the mean scores and standard deviations obtained through the cross-validation process.  
    
    """
    train_scores, test_scores = validation_curve(
    model, X_train, y_train, param_name=param, param_range=param_grid,
    scoring = scoring, cv = cv)
    
    train_mean = np.mean(train_scores, axis = 1)
    train_std = np.std(train_scores, axis = 1)

    test_mean = np.mean(test_scores, axis = 1)
    test_std = np.std(test_scores, axis = 1)
    
    plt.subplots(1, figsize = (width, height))
    plt.plot(param_grid, train_mean, label = 'Training score', color = 'black')
    plt.plot(param_grid, test_mean, label = 'Validation score', color = 'brown')

    plt.fill_between(param_grid, train_mean - train_std, train_mean + train_std, color = 'blue', alpha = 0.2)
    plt.fill_between(param_grid, test_mean - test_std, test_mean + test_std, color = 'darkblue', alpha = 0.2)
 
    plt.title(title, fontsize = 14, fontweight = 'bold')
    plt.xlabel('Param Range')
    plt.ylabel(str(scoring) +' ' + 'score')
    plt.tight_layout()
    plt.legend(loc = 'best')
    plt.show()

def train_val_test(data = None, class_labels = None, train = 0.6, val = 0.2, shuffle = True, random_state = None):
    
    """
    Function that accepts a Pandas dataframe and will return a training, validation, and test set. 
    """
    
    if shuffle:
        data = data.sample(frac = 1, random_state = random_state)
    else:
        pass
    
    X = data.drop(class_labels, axis = 1)
    y = pd.DataFrame(data[class_labels])
    
    split = train + val
    X_train, X_val, X_test = np.split(X, [int(train*len(X)), int(split*len(X))])
    y_train, y_val, y_test = np.split(y, [int(train*len(y)), int(split*len(y))])
    
    return X_train, y_train, X_val, y_val, X_test, y_test

def pca_scree_plot(data = None, n_components = None, width = 16, height = 10, legend_size = 12, scale_data = False, title = 'PCA Scree Plot'):

    """
    Function that accepts data, and number of principal components to be analysed. It will produce a scree plot of the cumulative variance explained.  
    """
    if scale_data:
        data = scale(data)
    else:
        pass

    pca = PCA(n_components = n_components)
    pca_model = pca.fit(data)
    var_exp = pca_model.explained_variance_ratio_.cumsum().round(4)*100
    
    fig, ax = plt.subplots(figsize = (width, height))
    ax.set_title(title, fontsize = 14, fontweight = 'bold')
    ax.set_xlabel('Principal Components', fontsize = 12)
    ax.set_ylabel('Variance Explained (%)', fontsize = 12)
    ax.axhspan(90, 100, alpha = 0.3, color = '#FF8C78', label = '90% - 95%')
    ax.axhspan(95, 100, alpha = 0.5, color = '#FF8C78', label = '95% - 99%')
    ax.axhspan(99, 100, alpha = 0.7, color = '#FF8C78', label = '99% - 100%')
    ax.legend(loc = 4, prop = {'size': legend_size})
    ax.plot(var_exp, marker = '.', markersize = 10);
    plt.show()            